### PR TITLE
snapshotstate: detect duplicated snapshot imports

### PR DESCRIPTION
**Pull request by mvo5**
This detects when someone tries to import a snapshot that is already available.

### CODE REVIEW

File: client/snapshot.go [hunk 1]
```diff
@@ -88,6 +90,21 @@ func (sh *Snapshot) IsValid() bool {
 	return !(sh == nil || sh.SetID == 0 || sh.Snap == "" || sh.Revision.Unset() || len(sh.SHA3_384) == 0 || sh.Time.IsZero())
 }
 
+// ContentHash returns a hash that can be used to identify the snapshot
+// but does not contain data like "time"
+func (sh *Snapshot) ContentHash() ([]byte, error) {
+	sh2 := *sh
+	sh2.SetID = 0
+	sh2.Time = time.Time{}
+	sh2.Auto = false
```
Comment (pedronis) [inline client/snapshot.go:99]: this is also annoying (that being auto or not is stored in the files) :/  anyway we need to discuss what to do about expiring snapshots vs imports

File: overlord/snapshotstate/backend/backend.go [hunk 2]
```diff
@@ -589,7 +593,43 @@ func writeOneSnapshotFile(targetPath string, tr io.Reader) error {
 	return nil
 }
 
-func unpackVerifySnapshotImport(r io.Reader, realSetID uint64) (snapNames []string, err error) {
+type DuplicatedSnapshotImportError struct {
+	setID uint64
+}
+
+func (e DuplicatedSnapshotImportError) Error() string {
```
Comment (stolowski) [inline overlord/snapshotstate/backend/backend.go:600]:
```suggestion
func (e *DuplicatedSnapshotImportError) Error() string {
```

File: overlord/snapshotstate/backend/backend.go [hunk 3]
```diff
@@ -589,7 +593,43 @@ func writeOneSnapshotFile(targetPath string, tr io.Reader) error {
...
+		return fmt.Errorf("cannot calculate snapshot set hashes: %v", err)
+	}
+
+	for setID, ss := range snapshotSetMap {
+		h, err := ss.ContentHash()
+		if err != nil {
+			return fmt.Errorf("cannot calculate content hash for %v: %v", setID, err)
+		}
+		if bytes.Equal(h, contentHash) {
+			return DuplicatedSnapshotImportError{setID}
```
Comment (stolowski) [inline overlord/snapshotstate/backend/backend.go:626]:
```suggestion
			return &DuplicatedSnapshotImportError{setID}
```

File: overlord/snapshotstate/backend/backend.go [hunk 4]
```diff
@@ -589,7 +593,43 @@ func writeOneSnapshotFile(targetPath string, tr io.Reader) error {
 	return nil
 }
 
-func unpackVerifySnapshotImport(r io.Reader, realSetID uint64) (snapNames []string, err error) {
+type DuplicatedSnapshotImportError struct {
```
Comment (stolowski) [inline overlord/snapshotstate/backend/backend.go:596]: Does it have to be exported?

Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:596]: A good question, I was thinking it might be useful in the clients but maybe YAGNI and all that?

File: overlord/snapshotstate/backend/backend.go (hunk 3 above)
Comment (mvo5) [reply overlord/snapshotstate/backend/backend.go:626]: No strong opinion, it just felt that a struct with a single uint64 is fine to use as a value (even the fact that it's a struct is a bit debatable, probably a `type DuplicatedSnapshotImportError uint64` would be fine but it feels slightly less clear. Anyway, I am happy either way I think. Maybe @pedronis has a stronger opinion?

File: overlord/snapshotstate/backend/backend.go [hunk 5]
```diff
@@ -565,7 +565,11 @@ func Import(ctx context.Context, id uint64, r io.Reader) (snapNames []string, er
 	defer tr.Cancel()
 
 	// Unpack and validate the streamed data
-	snapNames, err = unpackVerifySnapshotImport(r, id)
+	//
+	// XXX: this will leak snapshot IDs, i.e. we allocate a new
+	// snapshot ID before but then we error here because of e.g.
+	// duplicated import attempts
```
Comment (stolowski) [inline overlord/snapshotstate/backend/backend.go:571]: I don't think this is a problem, it's is just an internal counter and these numbers generally shouldn't bother end users. FWTW be do create automatic snapshots that expire after a while behind user's back, so "gaps" in the numbering will be there.

File: overlord/snapshotstate/backend/backend.go (hunk 3 above)
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:626]: Fair point, I think that's fine, I thought we use pointers for errors but we do cases similar to this already. So, +1

File: overlord/snapshotstate/backend/backend.go (hunk 5 above)
Comment (pedronis) [reply overlord/snapshotstate/backend/backend.go:676]: instead of failing we also have the option to change import to return the old id and return that to the client, but we also need to reset any expiration if the old snapshot was automatic (ideally we should stop marking if a snapshot is automatic in the zip but just carry that info the state where we carry expirations as well)

File: overlord/snapshotstate/backend/backend.go [hunk 6]
```diff
@@ -857,11 +924,36 @@ func (se *SnapshotExport) Close() {
...
 	defer tw.Close()
+
+	// export contentHash as early.json
+	h, err := json.Marshal(earlyJSON{se.contentHash})
+	if err != nil {
+		return err
+	}
+	hdr := &tar.Header{
+		Typeflag: tar.TypeReg,
+		Name:     "early.json",
```
Comment (pedronis) [inline overlord/snapshotstate/backend/backend.go:944]: content.json perhaps?  another possible name if we want to convey that it comes early would be preview.json

**Review by pedronis — COMMENTED** [review]
couple high-level remarks

File: overlord/snapshotstate/backend/backend.go (hunk 5 above)
Comment (stolowski) [reply overlord/snapshotstate/backend/backend.go:666]: Done. I kept DuplicatedSnapshotImportError internally between backend and snapshotstate but it is not surfaced anymore; an alternative approach would be to return actual set id but slightly annoying (less readable) as it goes through 2 helpers (unpackVerifySnapshotImport and checkDuplicatedSnapshotSetWithContentHash).

File: overlord/snapshotstate/snapshotstate.go [hunk 7]
```diff
@@ -333,6 +333,19 @@ func Import(ctx context.Context, st *state.State, r io.Reader) (setID uint64, sn
 	}
 	snapNames, err = backendImport(ctx, setID, r)
 	if err != nil {
+		if dupErr, ok := err.(backend.DuplicatedSnapshotImportError); ok {
+			st.Lock()
+			defer st.Unlock()
+			// trying to import identical snapshot; instead return set ID of
+			// the existing one and reset its expiry time.
+			// XXX: at the moment expiry-time is the only attribute so we can
```
Comment (mvo5) [inline overlord/snapshotstate/snapshotstate.go:341]: Isn't the auto flag also something we store into the state?

**Review by mvo5 — COMMENTED** [review]
Thanks for these changes Pawel, looking good!

File: client/snapshot.go [hunk 8]
```diff
@@ -91,6 +93,21 @@ func (sh *Snapshot) IsValid() bool {
 	return !(sh == nil || sh.SetID == 0 || sh.Snap == "" || sh.Revision.Unset() || len(sh.SHA3_384) == 0 || sh.Time.IsZero())
 }
 
+// ContentHash returns a hash that can be used to identify the snapshot
```
Comment (pedronis) [inline client/snapshot.go:96]: ...the snapshot by its content, leaving out metadata like "time" or "set-id".

File: client/snapshot.go [hunk 9]
```diff
@@ -120,6 +137,29 @@ func (ss SnapshotSet) Size() int64 {
 	return sum
 }
 
+type bySnap []*Snapshot
+
+func (ss bySnap) Len() int           { return len(ss) }
+func (ss bySnap) Swap(i, j int)      { ss[i], ss[j] = ss[j], ss[i] }
+func (ss bySnap) Less(i, j int) bool { return ss[i].Snap < ss[j].Snap }
+
+// ContentHash returns a hash that can be used to identify the SnapshotSet
```
Comment (pedronis) [inline client/snapshot.go:146]: ....by its content.

File: overlord/snapshotstate/snapshotstate.go [hunk 10]
```diff
@@ -333,6 +333,19 @@ func Import(ctx context.Context, st *state.State, r io.Reader) (setID uint64, sn
 	}
 	snapNames, err = backendImport(ctx, setID, r)
 	if err != nil {
+		if dupErr, ok := err.(backend.DuplicatedSnapshotImportError); ok {
```
Comment (pedronis) [inline overlord/snapshotstate/snapshotstate.go:336]: it's very unlikely but given we don't hold any kind of lock there's a chance that the snapshot is getting forgotten around here, before we get a chance to take the state lock.

It might be that easiest solution is really to have some kind of snapshot manager local lock to use around the listing of what is on disk and until the end of initialisation of some operations...

**Review by pedronis — COMMENTED** [review]
maybe we have a raciness issue

File: overlord/snapshotstate/snapshotstate.go (hunk 7 above)
Comment (stolowski) [reply overlord/snapshotstate/snapshotstate.go:341]: No, it was stored inside snapshot json, but in the state we only keep expiry-time for automatic snaphots.

**Comment by pedronis** [comment]
@mvo5 @stolowski looking at this and the other code more I think we have two issues, related partly to the fact that export/import are synchronous operations without matching changes/tasks:

* one could trigger a forget or expiration could happen for a snapshot while it's being exported,  the issue is that while we don't have clear infrastructure for that conceptually an export like a restore should conflict with forget and the other way around

* while we decide if an import already exists we also need to stop forgetting or auto-expiration to happen

we need to be careful though because auto-expiration code happens in Ensure so we shouldn't do something that blocks it completely we can have it not run or or ignore snapshots though

**Comment by stolowski** [comment]
@pedronis The remark about conflicts has a followup PR now, the other minor things have been addressed here.

**Review by pedronis — APPROVED** [review]
apart the raciness issues (that will be addressed in the follow up) this looks good to me, thanks

File: client/snapshot_test.go [hunk 11]
```diff
@@ -217,3 +218,74 @@ func (cs *clientSuite) TestClientSnapshotImport(c *check.C) {
...
+	c.Assert(err, check.IsNil)
+	c.Check(h1, check.DeepEquals, h1_1)
+
+	// same except set means same hash
+	h2, err := sh2.ContentHash()
+	c.Assert(err, check.IsNil)
+	c.Check(h1, check.DeepEquals, h2)
+
+	// sh3 is actually different
+	sh3 := &client.Snapshot{SetID: 1, Time: now, Snap: "other-snap", Revision: revno, SHA3_384: sums}
```
Comment (anonymouse64) [inline client/snapshot_test.go:249]: maybe too paranoid, but should we also have a case where the sums (i.e. the SHA3_384 key) is different and everything else (i.e. the Snap, the Revision, etc.) is the same?

File: overlord/snapshotstate/backend/backend_test.go [hunk 12]
```diff
@@ -989,6 +990,35 @@ func (s *snapshotSuite) TestImportCheckErorr(c *check.C) {
...
+
+	buf := bytes.NewBuffer(nil)
+	c.Assert(export.StreamTo(buf), check.IsNil)
+	c.Check(buf.Len(), check.Equals, int(export.Size()))
+
+	// now import it
+	_, err = backend.Import(ctx, 123, buf)
+	dupErr, ok := err.(backend.DuplicatedSnapshotImportError)
+	c.Assert(ok, check.Equals, true)
+	c.Check(dupErr.SetID, check.Equals, shID)
```
Comment (anonymouse64) [inline overlord/snapshotstate/backend/backend_test.go:1019]: nitpick but my slightly preferred way to do this sort of comparison is:

```go
c.Assert(dupErr,DeepEquals,backend.DuplicatedSnapShotImportError{SetID: shID})
```

this ensures that if new fields are added that we should be checking the test will start failing if we don't update the test in conjunction too

File: tests/main/snapshot-basic/task.yaml [hunk 13]
```diff
@@ -81,6 +72,22 @@ execute: |
...
 
+    echo "Check that import works"
+    RESTORE_ID=$( snap import-snapshot "${SET_ID}_export.snapshot" | head -n1 | cut -f2 -d'#' )
+    echo "And is valid"
+    snap check-snapshot "$RESTORE_ID"
+
+    NUM_SAVED=$(snap saved | wc -l)
+
+    echo "Import again doesn't create a duplicated snapshot"
+    IMPORTED_ID=$( snap import-snapshot "${SET_ID}_export.snapshot" | head -n1 | cut -f2 -d'#' )
```
Comment (anonymouse64) [inline tests/main/snapshot-basic/task.yaml:83]: could this use `SET_ID}_export-2.snapshot` for extra robustness? It seems to me that since we exported the same snapshot to two different files (which should be the same file too), we should get a duplicated set ID if we use the second file same as using the first file

File: overlord/snapshotstate/snapshotstate.go [hunk 14]
```diff
@@ -317,12 +317,6 @@ func List(ctx context.Context, st *state.State, setID uint64, snapNames []string
 	return sets, nil
 }
 
-// XXX: Something needs to cleanup incomplete imports. This is conceptually
```
Comment (anonymouse64) [inline overlord/snapshotstate/snapshotstate.go:320]: was this just done in a different pr or is this no longer a concern?

**Review by anonymouse64 — APPROVED** [review]
lgtm, a couple little nitpicks and improvements which I think are fine for followups

also a clarification on a deleted todo about whether that was done elsewhere or is no longer a concern since afaict this pr does not address that todo

File: overlord/snapshotstate/snapshotstate.go (hunk 14 above)
Comment (stolowski) [reply overlord/snapshotstate/snapshotstate.go:320]: Yes, was addressed before but we forgot to remove this comment.

File: tests/main/snapshot-basic/task.yaml (hunk 13 above)
Comment (stolowski) [reply tests/main/snapshot-basic/task.yaml:83]: Yes, makes sense to do it that way, thanks.

File: client/snapshot_test.go (hunk 11 above)
Comment (stolowski) [reply client/snapshot_test.go:249]: Sure, added.
